src/Log/Engine/BaseLog.php

```diff
@@ -90,6 +93,66 @@ public function scopes()
      */
     protected function _format(string $message, array $context = []): string
     {
-        return $message;
+        if (strpos($message, '{') === false && strpos($message, '}') === false) {
+            return $message;
+        }
+
+        preg_match_all('/\{([a-z][a-z0-9-_]*)\}/i', $message, $matches);
+        if (empty($matches)) {
+            return $message;
+        }
+
+        $placeholders = array_intersect($matches[1], array_keys($context));
+        $replacements = [];
+
+        foreach ($placeholders as $key) {
+            $value = $context[$key];
+
+            if (is_scalar($value)) {
+                $replacements['{' . $key . '}'] = (string)$value;
+                continue;
+            }
+
+            if (is_array($value)) {
+                $replacements['{' . $key . '}'] = print_r($value, true);
```

Comment:
Could this have the same memory exhaustion error that we had before? Should we use json_encode here instead?

Reply:
It potentially could. There are few other `print_r()` uses too, should I replace them all with `json_encode()`?

Reply:
I think it depends on how print_r() is being used. I personally find its multiline output annoying in logs especially when they are forwarded via syslog. JSON in that scenario is easier to manipulate downstream.

Reply:
I don't like its multi line output either. I used it just cause that's what was done in the 3.x code. I'll replace it with json_encode().

Reply:
Done